core/store/src/trie/mem/parallel_loader.rs

```diff
@@ -0,0 +1,420 @@
+            // (The first 8 bytes of the key is the ShardUId).
+            let key = NibbleSlice::new(&key[8..]).mid(subtree_to_load.prefix.len());
+            let value = FlatStateValue::try_from_slice(&value).map_err(|err| {
+                FlatStorageError::StorageInternalError(format!(
+                    "invalid FlatState value format: {err}"
+                ))
+            })?;
+            recon.add_leaf(key, value);
+        }
+        Ok(recon.finalize().unwrap())
+    }
+
+    /// This implements stage 2 and 3, loading the subtrees in parallel an then constructing the
+    /// final trie.
+    fn load_in_parallel(
+        &self,
+        plan: PartialTrieLoadingPlan,
+        name: String,
+    ) -> Result<(STArena, MemTrieNodeId), StorageError> {
+        let arena = ConcurrentArena::new();
+
+        // A bit of an awkward Rayon dance. We run a multi-threaded fold; the fold state contains
+        // both a sparse vector of the loading results as well as the arena used for the thread.
+        // We need to collect both in the end, so fold is the only suitable method.
+        let (roots, threads): (
+            Vec<Vec<Result<(usize, MemTrieNodeId), StorageError>>>,
+            Vec<ConcurrentArenaForThread>,
+        ) = plan
+            .subtrees_to_load
+            .into_par_iter()
+            .enumerate()
+            .fold(|| -> (Vec<Result<(usize, MemTrieNodeId), StorageError>>, ConcurrentArenaForThread) {
+                (Vec::new(), arena.for_thread())
+            }, |(mut roots, mut arena), (i, prefix)| {
+                roots.push(self.load_one_subtree(&prefix, &mut arena).map(|root| (i, root)));
+                (roots, arena)
+            })
+            .unzip();
+
+        let mut roots = roots.into_iter().flatten().collect::<Result<Vec<_>, _>>()?;
+        roots.sort_by_key(|(i, _)| *i);
+        let roots = roots.into_iter().map(|(_, root)| root).collect::<Vec<_>>();
+
+        let mut arena = arena.to_single_threaded(name, threads);
+        let root = plan.root.to_node(&mut arena, &roots);
+        Ok((arena, root))
+    }
+}
+
+/// Specifies exactly what to do to create a node in the final trie.
+#[derive(Debug)]
+enum TrieLoadingPlanNode {
+    // The first three cases correspond exactly to the trie structure.
+    Branch { children: Vec<(u8, Box<TrieLoadingPlanNode>)>, value: Option<FlatStateValue> },
+    Extension { extension: Box<[u8]>, child: Box<TrieLoadingPlanNode> },
+    Leaf { extension: Box<[u8]>, value: FlatStateValue },
+    // This means this trie node is whatever loading this subtree yields.
+    Load { subtree_id: usize },
+}
+
+impl TrieLoadingPlanNode {
+    /// This implements the construction part of stage 3, where we convert a plan node to
+    /// a memtrie node. The `subtree_roots` is the parallel loading results.
+    fn to_node(self, arena: &mut impl Arena, subtree_roots: &[MemTrieNodeId]) -> MemTrieNodeId {
```

Comment:
To clarify, is this the exact thing allowing to get rid of delayed hash computation?

Reply:
The purpose of delaying hash computation is to speed up loading, not that we have to delay it. Computing hash right away is always possible when constructing a memtrie node.